### PR TITLE
Implement various services for debug monitor usage.

### PR DESCRIPTION
**Pull request by SciresM**
This PR adds ro:dmnt, and fully fleshes out ns:dev, pm:dmnt, pm:shell. It also adds ldr:ro, since that was apparently missing..

### CODE REVIEW

File: nx/include/switch/services/ns.h [hunk 1]
```diff
@@ -34,6 +35,28 @@ typedef struct
...
+typedef enum {
+    NsShellEvent_None = 0,
+    NsShellEvent_Exit = 1,
+    NsShellEvent_Start = 2,
+    NsShellEvent_Crash = 3,
+    NsShellEvent_Debug = 4,
+} NsShellEvent;
+
+typedef struct {
+    NsShellEvent event : 32;
```
Comment (fincs) [inline nx/include/switch/services/ns.h]: The `: 32` isn't necessary, because on AArch64 enums are at least 32-bit.

File: nx/source/services/ro.c [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+#include "arm/atomics.h"
+#include "kernel/ipc.h"
+#include "kernel/detect.h"
+#include "services/ro.h"
+#include "services/sm.h"
+
+static Service g_roSrv, g_dmntSrv;
+static u64 g_roRefCnt, g_dmntRefCnt;
+
+Result _ldrRoInitialize(void);
```
Comment (fincs) [inline nx/source/services/ro.c]: Should be declared as static

File: nx/source/services/ro.c [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+        serviceIpcParse(&g_roSrv, &r, sizeof(*resp));
+        resp = r.Raw;
+
+        rc = resp->result;
+    }
+
+    return rc;
+}
+
+Result _ldrRoInitialize(void) {
```
Comment (yellows8) [inline nx/source/services/ro.c]: static

File: nx/include/switch/services/ns.h [hunk 4]
```diff
@@ -65,5 +87,15 @@ Result nsvmGetSafeSystemVersion(u16 *out);
 Result nsdevInitialize(void);
 void nsdevExit(void);
 
+Result nsdevLaunchProgram(u64* out_pid, const NsLaunchProperties* properties, u32 flags);
 Result nsdevTerminateProcess(u64 pid);
 Result nsdevTerminateProgram(u64 tid);
+Result nsdevGetShellEvent(Event* out);
```
Comment (yellows8) [inline nx/include/switch/services/ns.h]: Add comment for autoclear.

File: nx/include/switch/services/pm.h [hunk 5]
```diff
@@ -18,14 +42,23 @@ void pminfoExit(void);
...
 Result pmdmntGetApplicationPid(u64* pid_out);
 Result pmdmntEnableDebugForApplication(Handle* handle_out);
+Result pmdmntDisableDebug(void);
 
 Result pminfoGetTitleId(u64* title_id_out, u64 pid);
 
 Result pmshellLaunchProcess(u32 launch_flags, u64 titleID, u64 storageID, u64 *pid);
+Result pmshellTerminateProcessByProcessId(u64 processID);
 Result pmshellTerminateProcessByTitleId(u64 titleID);
+Result pmshellGetProcessEvent(Event* out);
```
Comment (yellows8) [inline nx/include/switch/services/pm.h]: Add comment for autoclear.

File: nx/source/services/ro.c (hunk 3 above)
Comment (fincs) [reply nx/source/services/ro.c]: static only needs to be placed in the prototype, not in the function definition. In fact you get a warning iirc.

**Pull request merged**